# Deep Learning Architectures: CNNs and RNNs

## 📚 Learning Objectives

By completing this notebook, you will:
- Describe CNN and RNN architectures and use cases
- Read and compare `model.summary()` outputs (training happens in notebooks 05-06)

## 🔗 Where this fits

**Builds on:** Unit 4, lesson 01 "Simple Perceptron" — one neuron becomes a layer, and layers become architectures with a shape.

**Used later in:** Course 08 (AIAT 122) — Units 2 and 3, which train the CNNs and RNNs only described here.

---


## 🎯 One contest result that redirected an industry

The ImageNet Large Scale Visual Recognition Challenge asks a model to name what
is in a photograph, over a thousand categories. In **2012** the entry from Alex
Krizhevsky, Ilya Sutskever and Geoffrey Hinton — a **convolutional neural
network**, now called AlexNet — finished with a top-5 error rate of **15.3%**.
The next best entry, built on the hand-engineered features that had dominated
computer vision for a decade, scored **26.2%**. That gap of nearly eleven points
ended one research programme and started another; within three years essentially
every serious vision system was a CNN.

The sequence side has an equally dated turning point. In **September 2016** Google
replaced the phrase-based statistical engine behind Google Translate with
**GNMT** — a stack of eight LSTM encoder layers and eight LSTM decoder layers with
attention. Google reported it reduced translation errors by **55-85%** on several
language pairs against the system it replaced, and put it into production first
for Chinese→English. The LSTM you are about to define below is the same cell type,
at a smaller size.

### What goes wrong without matching the architecture to the data

A `Dense` layer treats its inputs as an unordered bag of numbers. Feed it an image
and you have thrown away the fact that neighbouring pixels form strokes and edges;
feed it a sequence and you have thrown away the order. Both are recoverable only
by making the network relearn, from data, a structure you already knew. That is
what a convolution and a recurrence *are*: prior knowledge about the data's shape,
built into the wiring so the network does not have to spend capacity discovering it.


# Deep Learning Architectures: CNNs and RNNs

**Unit:** Unit 4: Neural Networks Fundamentals  

## 📚 Learning Objectives

By completing this notebook, you will:
- Understand Convolutional Neural Networks (CNNs) structure
- Learn Recurrent Neural Networks (RNNs), LSTM, and GRU
- Compare CNNs and RNNs for different applications
- Define a CNN architecture and inspect its `model.summary()`
- Define RNN, LSTM, and GRU architectures and compare their parameter counts

**Note:** this notebook only *builds* the architectures — actual training on images
and sequences happens in notebooks 05 and 06.

---

## 📥 Inputs & 📤 Outputs

**Inputs:** What we use in this notebook

- Libraries and concepts as introduced in this notebook; see prerequisites and code comments.

**Outputs:** What you'll see when you run the cells

- Printed results, figures, and summaries as shown when you run the cells.

---

In [1]:
# Map of the deep-learning architecture zoo: CNNs for spatial data, RNN/LSTM/GRU for
# sequences. Choosing the architecture that matches the data's structure is half the job.
import numpy as np
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, LSTM, GRU, SimpleRNN

print("=== Deep Learning Architectures ===")
print("\n1. CNN (Convolutional Neural Network):")
print("   - Best for: Images, spatial data")
print("   - Components: Convolution, Pooling, Fully Connected")
print("   - Applications: Image classification, object detection")

print("\n2. RNN (Recurrent Neural Network):")
print("   - Best for: Sequential data, time series")
print("   - Components: Recurrent layers, memory")
print("   - Applications: Text, speech, time series")

print("\n3. LSTM (Long Short-Term Memory):")
print("   - Advanced RNN with long-term memory")
print("   - Solves vanishing gradient problem")

print("\n4. GRU (Gated Recurrent Unit):")
print("   - Simpler than LSTM, similar performance")

=== Deep Learning Architectures ===

1. CNN (Convolutional Neural Network):
   - Best for: Images, spatial data
   - Components: Convolution, Pooling, Fully Connected
   - Applications: Image classification, object detection

2. RNN (Recurrent Neural Network):
   - Best for: Sequential data, time series
   - Components: Recurrent layers, memory
   - Applications: Text, speech, time series

3. LSTM (Long Short-Term Memory):
   - Advanced RNN with long-term memory
   - Solves vanishing gradient problem

4. GRU (Gated Recurrent Unit):
   - Simpler than LSTM, similar performance


## Example: CNN for Image Classification


In [2]:
# Build (not train) a small CNN and read its summary: watch the spatial size shrink
# through conv/pooling while the channel count grows — the CNN signature.
# Simple CNN architecture example
print("=== CNN Architecture Example ===")
print("\nCNN Structure:")
print("Input → Conv2D → MaxPooling → Conv2D → MaxPooling → Flatten → Dense → Output")

# Example CNN model (simplified for demonstration)
cnn_model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    MaxPooling2D((2, 2)),
    Conv2D(64, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Flatten(), Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

print("\nModel Summary:")
cnn_model.summary()
print("\n✅ CNN is designed for spatial patterns in images")

=== CNN Architecture Example ===

CNN Structure:
Input → Conv2D → MaxPooling → Conv2D → MaxPooling → Flatten → Dense → Output

Model Summary:


/Users/abdullah/venvs/ai-diploma-tf/lib/python3.13/site-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 26, 26, 32)     │           320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 13, 13, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 11, 11, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 5, 5, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 1600)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 64)             │       102,464 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │           650 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 121,930 (476.29 KB)

 Trainable params: 121,930 (476.29 KB)

 Non-trainable params: 0 (0.00 B)


✅ CNN is designed for spatial patterns in images


## Example: RNN/LSTM for Sequential Data


In [3]:
# Same exercise for sequences: an LSTM reads 10 timesteps one by one and keeps a
# memory — its parameter count comes from the gates, visible in the summary.
# Simple RNN/LSTM architecture example
print("=== RNN/LSTM Architecture Example ===")
print("\nRNN Structure:")
print("Input sequence → RNN/LSTM layers → Dense → Output")

# Example LSTM model for sequence prediction
lstm_model = Sequential([
    LSTM(50, activation='relu', input_shape=(10, 1)),  # 10 timesteps, 1 feature
    Dense(1)
])

print("\nModel Summary:")
lstm_model.summary()
print("\n✅ LSTM/RNN is designed for temporal/sequential patterns")

=== RNN/LSTM Architecture Example ===

RNN Structure:
Input sequence → RNN/LSTM layers → Dense → Output

Model Summary:


/Users/abdullah/venvs/ai-diploma-tf/lib/python3.13/site-packages/keras/src/layers/rnn/rnn.py:199: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ lstm (LSTM)                     │ (None, 50)             │        10,400 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            51 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 10,451 (40.82 KB)

 Trainable params: 10,451 (40.82 KB)

 Non-trainable params: 0 (0.00 B)


✅ LSTM/RNN is designed for temporal/sequential patterns


## 📊 Read the two summaries side by side

Both models were only *built*, never trained — but their parameter counts already
tell you what each architecture assumes.

**In the CNN summary**, follow two numbers down the table: the spatial size shrinks
(28×28 → 26×26 → 13×13 → …) while the channel count grows (1 → 32 → 64). That is
the CNN signature — trade *where* something is for *what* it is. Notice also that
the first `Conv2D` has only **320** parameters: 32 filters × (3×3 kernel + 1 bias).
Those 320 numbers are reused at every one of the 676 positions in the image. A
`Dense` layer connecting 784 pixels to 32 units would need 25,120.

**In the LSTM summary**, the single `LSTM(50)` layer holds **10,400** parameters
for an input of 10 timesteps × 1 feature. Work out why: an LSTM has four internal
gates, each with weights over the input (1), the previous hidden state (50) and a
bias — 4 × (1 + 50 + 1) × 50 = 10,400. **The gates are where the parameters go**,
and that is the price of memory that survives many timesteps. Notebook 06 trains
all three cell types and shows what the price buys.


## 💬 Discuss

1. The CNN's first layer reuses 320 parameters across every position in the image.
   That assumption — a useful pattern is useful *anywhere* in the input — is called
   translation invariance. **Name a kind of data where it is false**, and say what
   a CNN would get wrong there.
2. An LSTM processes timesteps one after another, so it cannot be parallelised
   across time; a transformer sees the whole sequence at once. Given the parameter
   count above, argue when you would still choose an LSTM in 2026 — think about
   sequence length, hardware, and the size of your dataset.
3. You have 5,000 labelled chest X-rays and a deadline. Build a CNN from scratch,
   or fine-tune a network pre-trained on everyday photographs? Commit to one, then
   name the result from notebook 05 that would make you switch.


## ⚠️ Where this breaks

- **This notebook builds architectures; it does not evaluate them.** A
  `model.summary()` proves the shapes are consistent and nothing else. Parameter
  counts are not quality, and the CNN below has never seen an image. Notebooks 05
  and 06 do the actual measuring, and one of them shows a *deeper* model failing
  to beat a shallower one.
- **A CNN assumes locality and translation invariance.** Both hold for photographs
  and hold badly for tabular data, where column order is arbitrary — shuffle the
  columns of a spreadsheet and nothing changes about the data, but a convolution
  over them becomes meaningless. Reaching for a CNN because the input "is a
  matrix" is a common and expensive mistake.
- **Pooling throws away position on purpose.** Great for "is there a cat";
  actively harmful when the answer depends on *where* something is, which is why
  detection and segmentation networks are built differently.
- **A SimpleRNN cannot hold long-range information.** Its single state vector is
  overwritten at every step, so gradients over long sequences vanish. LSTM and GRU
  exist as fixes, and they are partial ones — Unit 4, notebook 06 measures a case
  where they barely beat the SimpleRNN because the useful history is short.
- **This framing predates the current architecture landscape.** "CNNs for images,
  RNNs for sequences" was the right map in 2016. Since then transformers have
  taken most sequence work, vision transformers compete with CNNs on large
  datasets, and transformer backbones now drive *image generation* too (Peebles &
  Xie, 2022, *Scalable Diffusion Models with Transformers*). Learn the two
  inductive biases here — locality and recurrence — as ideas, then expect the
  default architecture to have moved by the time you deploy.
- **Bigger is not the answer to a wrong bias.** If the architecture encodes the
  wrong assumption about your data, adding layers spends compute making the wrong
  assumption more confidently.


## 📚 References

1. LeCun, Y., Bottou, L., Bengio, Y., & Haffner, P. (1998). *Gradient-Based Learning Applied to Document Recognition*. Proceedings of the IEEE, 86(11), 2278–2324.
2. Hochreiter, S., & Schmidhuber, J. (1997). *Long Short-Term Memory*. Neural Computation, 9(8), 1735–1780.
3. Krizhevsky, A., Sutskever, I., & Hinton, G. E. (2012). *ImageNet Classification with Deep Convolutional Neural Networks*. NeurIPS.
4. Dosovitskiy, A., Beyer, L., Kolesnikov, A., et al. (2021). *An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale* (ViT). ICLR. <https://arxiv.org/abs/2010.11929>